doc/developer/arrangements.md

```diff
@@ -0,0 +1,129 @@
+# Understanding what Costs in Materialize
+
+Materialize builds dataflow graphs that represents queries.
+The operators in these graphs can either be *stateless* (as in a Map, Filter, or Project operator) or *stateful*.
+Stateful operators hold on to arrangements of records, which are the main contributor to memory footprint in Materialize.
+Arrangements can be shared between operators that need the same data, which can ameliorate some of the cost.
+
+Let's talk through what an arrangement is, which operators require arrangements, and what we can do about them.
+
+## What is an "Arrangement"?
+
+An arrangement is an alternative representation of a stream of update triples `(data, time, diff)`.
+In stream form, the triple pass by and are gone.
+An arrangement records these triples, and indexes them by `data`.
+With some caveats about "logical compaction", the index maintains the *current* accumulation of `diff` for each `(data, time)`.
+This means that the size of an arrangement should be roughly the current size of the accumulated updates, where "zero" accumulations contribute nothing.
+An arrangement undergoes "physical compaction" lazily, which means that the actual size of an arrangement can exceed the logical size, for what should be a bounded amount of time.
+
+An arrangement only costs proportional to the number of distinct `(data, time)` pairs.
+This number can be small even if the number of records is large, if we were able to reduce `data` to few values.
+For example, if we want to histogram taxi rides by number of occupants and fare amount, there may be few distinct pairs despite many records.
+
+Arrangements are actually indexed assuming that `data` has a `(key, val)` structure.
+The index is by `key`, which will be important in understanding when arrangements can be shared and reused.
+Materialize will attempt to choose a `key` to ensure that the data are well distributed.
+For example, if a collection has a primary key, Materialize will default to this key for an arrangement.
+Each of `key` and `val` can be `()`, corresponding to "no key" and "the key is the value", respectively.
```

Comment:
I get what you're saying, but it'd be slightly clearer to say something like "if we don't know the primary key, we use the key as the value"